src/mono/mono/mini/aot-compiler.c

```diff
@@ -14473,11 +14699,59 @@ mono_compile_assembly (MonoAssembly *ass, guint32 opts, const char *aot_options,
 	}
 #endif
 
-	if (acfg->aot_opts.direct_pinvoke && !acfg->aot_opts.static_link) {
-		aot_printerrf (acfg, "The 'direct-pinvoke' AOT option also requires the 'static' AOT option.\n");
+	if ((acfg->aot_opts.direct_pinvoke || acfg->aot_opts.direct_pinvokes || acfg->aot_opts.direct_pinvoke_lists) && !acfg->aot_opts.static_link) {
+		aot_printerrf (acfg, "The 'direct-pinvoke' flag, 'direct-pinvokes', and 'direct-pinvoke-lists' AOT options also require the 'static' AOT option.\n");
+		return 1;
+	}
+	if (acfg->aot_opts.direct_pinvoke && (acfg->aot_opts.direct_pinvokes || acfg->aot_opts.direct_pinvoke_lists)) {
+		aot_printerrf (acfg, "The 'direct-pinvoke' flag trumps specified 'direct-pinvokes' and 'direct-pinvoke-lists' arguments. Unset either the flag or the specific direct pinvoke arguments.\n");
 		return 1;
 	}
 
+	gboolean added_direct_pinvoke = TRUE;
+	if (acfg->aot_opts.direct_pinvokes) {
+		GList *l;
+
+		for (l = acfg->aot_opts.direct_pinvokes; l; l = l->next) {
+			process_specified_direct_pinvokes (acfg, (char*)l->data);
+			if (!added_direct_pinvoke)
+				return 1;
+		}
+	}
+	if (acfg->aot_opts.direct_pinvoke_lists) {
+		GList *l;
+
+		for (l = acfg->aot_opts.direct_pinvoke_lists; l; l = l->next) {
+			char *direct_pinvoke_list = (char*)l->data;
+			FILE *direct_pinvoke_list_file = fopen (direct_pinvoke_list, "r");
+			if (!direct_pinvoke_list_file) {
+				aot_printerrf (acfg, "Failed to open the provided 'direct-pinvoke-list' '%s', fopen failed.\n", direct_pinvoke_list);
+				return 1;
+			}
+
+			char *line = NULL;
+			size_t line_len = 0;
+			while (getline (&line, &line_len, direct_pinvoke_list_file) != -1 && added_direct_pinvoke) {
```

Comment:
Look into use fgets with a fixed stack buffer, not sure it is available on all platforms needed.

Reply:
Replaced with `fgets` using a style similar to https://github.com/dotnet/runtime/blob/7bbee52b2aaa22f3116b96aedb6a758a2d5bd63b/src/mono/mono/mini/aot-compiler.c#L14147-L14191, but it feels like there are some things that are wrong in that one. (i.e. fgets will read the `'\n'` char, so `strlen (line)` will be the length including `'\n'`, so the logic to replace the newline char with null terminating char isn't hit. Moreover the provided `(int)(fileLength - offset)` means that `(int)(fileLength - offset) - 1` characters are read, so it seems like the last char will be omitted. )

There is one dynamic allocation for the buffer `dynamic_pinvokes` to read the file, and a pointer to each direct pinvoke read will be passed into `process_specified_direct_pinvokes`. Do we want to avoid that dynamic allocation based on the file size?

And as suggested, I drop the newline char (replace it with null terminating), call `g_strstrip`, and skip if the direct pinvoke is 0 length string after the cleans.

Reply:
see, https://github.com/dotnet/runtime/pull/79721#discussion_r1065558818